app/src/main/java/com/example/sharewhatyoucanproject/donor/DonorViewModel.kt

```diff
@@ -0,0 +1,126 @@
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import com.example.sharewhatyoucanproject.utils.showToast
+import com.google.android.gms.location.LocationCallback
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationResult
+import com.google.android.gms.location.LocationServices
+import com.google.firebase.firestore.FieldValue
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.GeoPoint
+import com.google.firebase.storage.StorageReference
+import com.google.firebase.storage.UploadTask
+import java.util.*
+
+class DonorViewModel(
+    val context: Context?,
+    var db: FirebaseFirestore?,
+    var storageReference: StorageReference?,
+) : ViewModel() {
+    private lateinit var uploadTask: UploadTask
+    private var locationRequest: LocationRequest = LocationRequest.create()
+
+    private val _currentLocation = MutableLiveData<GeoPoint>()
+    val currentLocation: LiveData<GeoPoint> get() = _currentLocation
+
+    private val _imageUrl = MutableLiveData<String>()
+    val imageUrl: LiveData<String> get() = _imageUrl
+
+    private val _saveData = MutableLiveData<String>()
+    val saveData: LiveData<String> get() = _saveData
+
+    var foodType: String = "Cooked Food"
+    var title = ""
+    var description = ""
+    var hourSet = ""
+    var filepath: Uri? = null
+
+    fun isDataCompleted(): Boolean {
+        return title.isNotEmpty() && description.isNotEmpty() && hourSet.isNotEmpty()
+    }
+
+    fun isImageNotSelected(): Boolean {
+        return filepath == null
+    }
+
+    fun isFoodNotValid(): Boolean {
+        val hours = Integer.parseInt(hourSet)
+        return foodType == "Cooked Food" && hours > 48 || foodType == "Groceries" && hours > 1460
+    }
+
+    fun saveData() {
+        val postMap: MutableMap<String, Any?> = HashMap()
+        postMap["imageUrl"] = imageUrl
+        postMap["date"] = FieldValue.serverTimestamp()
+        postMap["title"] = title
+        postMap["description"] = description
+        postMap["status"] = 0
+        postMap["type"] = foodType
+        postMap["location"] = currentLocation.value
+        db?.collection("posts")
+            ?.add(postMap)
+            ?.addOnCompleteListener { task ->
+                if (task.isSuccessful) {
+                    _saveData.value = "success"
+                } else {
+                    _saveData.value = "Failed ${task.exception}"
+                }
+            }
+    }
+
+    fun uploadImage() {
+        storageReference?.let {
+            val ref = it.child("images/" + UUID.randomUUID())
+            uploadTask = ref.putFile(filepath!!)
+            uploadTask.continueWithTask { task ->
+                if (!task.isSuccessful) {
+                    val excep = task.exception
+                    if (excep != null) {
+                        throw excep
```

Comment:
this will crash the app. Instead we usually want to log known exceptions to a crash reporting service like Crashlytics. You do not have to fix this as part of this PR, I recommend making a separate PR to integrate Crashlytics or another crash reporting service.